### PR TITLE
Add more tests for catching bad inputs.

### PR DESCRIPTION
**Pull request by maljub01**
- Make sure badly constructed oneof fields are handled properly.
- Add an additional test case for a non-empty proto extension.
- Add logic to make sure that the HashProto method recovers properly if
  the proto library panics (See golang/protobuf/issues/478 for an
  example).

  This case is included in the malformed oneof field tests as
  `&Singleton{Singleton: (*Singleton_TheString)(nil)}`.

### CODE REVIEW

File: object_hasher.go [hunk 1]
```diff
@@ -54,7 +62,7 @@ func (hasher *objectHasher) HashProto(pb proto.Message) ([]byte, error) {
 	// Make sure the proto itself is actually valid (ie. can be marshalled).
 	// If this fails, it probably means there are unset required fields or invalid
 	// values.
-	if _, err := proto.Marshal(pb); err != nil {
+	if _, err = proto.Marshal(pb); err != nil {
```
Comment (puellanivis) [inline object_hasher.go]: There is no reason to remove the shadowing of the `err` value here.

File: object_hasher.go [hunk 2]
```diff
@@ -40,7 +40,15 @@ type objectHasher struct {
 }
 
 // `HashProto` returns the object hash of a given protocol buffer message.
-func (hasher *objectHasher) HashProto(pb proto.Message) ([]byte, error) {
+func (hasher *objectHasher) HashProto(pb proto.Message) (h []byte, err error) {
+	// Ensure that we can recover if the proto library panics.
+	// See: https://github.com/golang/protobuf/issues/478
+	defer func() {
```
Comment (puellanivis) [inline object_hasher.go]: Swallowing panics is almost never the ideal solution. Your own issue report demonstrates that a swallowed panic by fmt.Println can itself result in an issue report that is complicated by a swallowed panic.

File: tests/badness_detection.go [hunk 3]
```diff
@@ -48,13 +58,30 @@ func TestBadness(t *testing.T, hashers ProtoHashers) {
...
+		}(),
+
 		// Create proto messages with unknown fields. That's bad.
 		forgetAllFields(&pb2_latest.PersonV1{Name: proto.String("Unbekannt")}),
 	}
 	for _, message := range badProtos {
 		_, err := hasher.HashProto(message)
 		if err == nil {
-			t.Errorf("Attempting to hash %T{ %+v} should have returned an error.", message, message)
+			t.Errorf("Attempting to hash %T{ %[1]v } should have returned an error.", message)
```
Comment (puellanivis) [inline tests/badness_detection.go]: Any optimization that this would afford complicates the simple straight-forward construction that was there before, and is unnecessary, as it is, itself, a test. And thus not performance critical.

File: object_hasher.go (hunk 2 above)
Comment (maljub01) [reply object_hasher.go]: This code is not swallowing the panic, but instead converting it into an error that's returned by the method to its caller.

Additionally, this [go blog post on panic and recover](https://blog.golang.org/defer-panic-and-recover) mentions that:

> The convention in the Go libraries is that even when a package uses panic internally, its external API still presents explicit error return values.

File: object_hasher.go (hunk 1 above)
Comment (maljub01) [reply object_hasher.go]: I agree that keeping the shadowing here would be harmless, but I prefer to avoid shadowing where it's not adding value.

File: tests/badness_detection.go (hunk 3 above)
Comment (maljub01) [reply tests/badness_detection.go]: This was not changed to optimize for performance, but rather for readability.

File: object_hasher.go (hunk 2 above)
Comment (puellanivis) [reply object_hasher.go]: So, the library itself is throwing the panics explicitly? Or are we catching language panics that are not actually internal to the package?

File: object_hasher.go (hunk 1 above)
Comment (puellanivis) [reply object_hasher.go]: My interest here is “if it ain’t broke, don’t fix it”. If you were working on this line anyways, then, meh.

But since the code is behaviorally equivalent…

File: tests/badness_detection.go (hunk 3 above)
Comment (puellanivis) [reply tests/badness_detection.go]: You also drop the “+” flag, so this isn’t just a readability change. And no, it does not really meet the bar of “better readability”… they are equivalently readable, and the use of indexed references is so uncommon, that it can be argued that using it where reduplication works just as well would not be increased readability.

I mean, whom are you targeting here for readability? Pedantic people who know all the language features, or the general programmer?

File: object_hasher.go (hunk 2 above)
Comment (maljub01) [reply object_hasher.go]: It's not the protobuf library itself but rather a dependency of it that explicitly throws the panics. That dependency being the "reflect" package.

See: https://github.com/golang/go/blob/master/src/reflect/value.go

It's unfortunate since the correct thing to do for the reflect package should've been to return errors instead, but given that it uses panics as part of its error reporting I think that this recover is the lesser of two evils.

Comment (maljub01) [reply object_hasher.go]: To be fair, "reflect" is also a heavily used dependency in the code of this library itself, so a defensive recover is probably also warranted on those grounds alone.

File: tests/badness_detection.go (hunk 3 above)
Comment (maljub01) [reply tests/badness_detection.go]: Dropping the "+" flag was a separate change made because it makes no difference when applied to protos. The output is the same before and after.

I agree that it's not a very common feature but it is straightforward enough that I don't think it would be confusing for someone who is familiar with other flags and encounters it for the first time.

File: object_hasher.go (hunk 2 above)
Comment (puellanivis) [reply object_hasher.go]: I just don’t see the merit of catching another library/package’s panics…
